AddHostObjectToScript Deep Dive Doc (#1750)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/how-to/hostobject.md

  * For example, there might be a native webcam API, which would require re-writing a large amount of code on the web side. Having the ability to call the native object's methods is quicker and more efficient than re-coding the object's methods on the web side of your app. In this case, your native-side code can pass the object to your app's web-side, JavaScript code, so that your JavaScript code can reuse the native API's methods.

Scenarios that may benefit from using host objects in script:

  * There is a keyboard API, and you want to call the `keyboardObject.showKeyboard` function from the web side.

  * JavaScript is sandboxed, limiting its ability on the native side. For example, if you need to access a file on the native side, you must use the native file system. If you have a native object exposed to JavaScript via `AddHostObjectToScript`, you can use it to manipulate files on the native file system.

> This article uses the WebView2 Win32 sample app to demonstrate some practical applications of `AddHostObjectToScript`. For more information about how to embed web content into native applications, see [Embed web content into native applications](/microsoft-edge/webview2/how-to/communicate-btwn-web-native).

Reviewer: Should this be a link?

Reviewer 2: It is. It works in the rendered version. Is it not working for you?

Reviewer: I meant the reference to our WebView2 Win32 Sample App. This is the repo: [WebView2APISample](https://github.com/MicrosoftEdge/WebView2Samples/tree/master/SampleApps/WebView2APISample)

Reviewer 2: yes, fixing

**Preview of the major steps in this article:**

1. Install Visual Studio, install git, clone the WebView2Samples repo, and open the solution.

> 1. Define the host object and implement `IDispatch` so that WebView2 recognizes the object.

Reviewer: So that WebView2 can project/add it to the web side.

Reviewer 2: fixed

> 1. Use `AddHostObjectToScript` to pass a method to the web.

Reviewer: To pass an object?

Reviewer 2: No idea ... question for @jasonstephen15 or @david-risney ?

Reviewer 3: Yes
```suggestion
1. Use `AddHostObjectToScript` to pass an object to the web.
```

Reviewer 2: fixed

    
    Implement `IDispatch` as described in [Type Libraries and the Object Description Language](/previous-versions/windows/desktop/automat/type-libraries-and-the-object-description-language).
    
    If the object you want to add to JavaScript doesn't already implement `IDispatch`, you need to write an `IDispatch` class wrapper for the object that you want to expose.

    There might be libraries to do this automatically.  To learn more about the steps that are needed to write an `IDispatch` class wrapper for the object that you want to expose, see [Automation](/previous-versions/windows/desktop/automat/automation-programming-reference).
    
1. Once the interface is defined in the IDL, save and compile the sample project in Visual Studio to create the translation lookaside buffer (TLB) file. You need to reference the TLB file from the C++ source code shown in the following section.

### Part 2B: Create the C++ object

In the WebView2 sample code, the `HostObjectSampleImpl.cpp` file takes the skeleton created in the COM IDL file and builds the C++ object.

> Implement all the functions that are defined in your object's interface, as we outlined in the IDL file. Be sure to implement the functions that are required by `IDispatch`.  The compiler will throw a warning if these functions aren't defined.

Reviewer: Have we verified this? I think not implementing a declared member will actually be an error.

Reviewer: @jasonstephen15

Next, we examine two specific properties that were defined in the IDL, to show how the IDL is related to the `.cpp` file.

1. In Visual Studio **Solution Explorer**, open **WebView2APISample** > **Source Files** > **HostObjectSampleImpl.cpp**.

1. Compare the property *declarations*, in `HostObjectSample.idl` ...
 
    ```csharp
    [propget] HRESULT Property([out, retval] BSTR* stringResult);
    [propput] HRESULT Property([in] BSTR stringValue);
    ```
    ... to the *implementation* of the object's properties, in `HostObjectSampleImpl.cpp`:

    ```cpp

>     STDMETHODIMP HostObjectSample::get_Property(BSTR* stringResult)

Reviewer: @jasonstephen15 do we have reference as to why `[propget] HRESULT Property` is `get_Property` in implementation?

    55                // calling RemoveHostObject first. This will replace the previous object
    56                // with the new object. In our case this is the same object and everything
    57                // is fine.
    58                CHECK_FAILURE(
    59                    m_webView->AddHostObjectToScript(L"sample", &remoteObjectAsVariant));
    60                remoteObjectAsVariant.pdispVal->Release();
    61                //! [AddHostObjectToScript]
    62            }
    ```

    Lines 31 to 46 show code specific to this sample app where we display HTML. Your app may have a different implementation of this code.  

1. Review line 33, which shows how to instantiate the COM object just defined in the IDL file. This is the object we will use later when we call `AddHostObjectToScript`. This gets us a pointer to the interface in `HostObjectSampleImpl.cpp`.

1. Review line 51, which casts the newly created COM object to an `IDispatch` type and then converts the object to a `VARIANT`. `VARIANT` types allow you to use data structures such as integers and arrays as well as more complex types such as `IDispatch`. 

>     For a full list of supported data types, see [VARIANT structure (oaidl.h) - Win32 apps | Microsoft Docs](/windows/win32/api/oaidl/ns-oaidl-variant).

Reviewer: @jasonstephen15  We should make sure to note that these are types supported by VARIANT itself, but we only support a subset of them, as described in `AddHostObjectToScript` [docs](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.1108.44#addhostobjecttoscript).

Reviewer: @vchapel pushed content update. Feel free to make wording/format changes as needed. Thanks.
